relayer/src/channel.rs

```diff
@@ -622,33 +640,51 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> Channel<ChainA, ChainB> {
         .map_err(|e| ChannelError::query_channel(channel_id.clone(), e))
     }
 
-    pub fn handshake_step(&mut self, state: State) -> Result<Vec<IbcEvent>, ChannelError> {
-        match (state, self.counterparty_state()?) {
-            (State::Init, State::Uninitialized) => Ok(vec![self.build_chan_open_try_and_send()?]),
-            (State::Init, State::Init) => Ok(vec![self.build_chan_open_try_and_send()?]),
-            (State::TryOpen, State::Init) => Ok(vec![self.build_chan_open_ack_and_send()?]),
-            (State::TryOpen, State::TryOpen) => Ok(vec![self.build_chan_open_ack_and_send()?]),
-            (State::Open, State::TryOpen) => Ok(vec![self.build_chan_open_confirm_and_send()?]),
-            _ => Ok(vec![]),
-        }
-    }
+    pub fn handshake_step(
+        &mut self,
+        state: State,
+    ) -> Result<(Option<IbcEvent>, Next), ChannelError> {
+        let res = match (state, self.counterparty_state()?) {
+            (State::Init, State::Uninitialized) => Some(self.build_chan_open_try_and_send()?),
+            (State::Init, State::Init) => Some(self.build_chan_open_try_and_send()?),
+            (State::TryOpen, State::Init) => Some(self.build_chan_open_ack_and_send()?),
+            (State::TryOpen, State::TryOpen) => Some(self.build_chan_open_ack_and_send()?),
+            (State::Open, State::TryOpen) => Some(self.build_chan_open_confirm_and_send()?),
+            (State::Open, State::Open) => return Ok((None, Next::Abort)),
+
+            // If the counterparty state is already Open but current state is TryOpen,
+            // return anyway as the final step is to be done by the counterparty worker.
+            (State::TryOpen, State::Open) => return Ok((None, Next::Abort)),
```

Comment:
I added a new exit condition for the connection/channel worker. This is required because if there is a race in the handshake step with an external party, such as `bootstrap_connection`, the worker may otherwise get stuck in a loop and keep processing new block events even when the handshake has completed.

Reply:
Good catch!